src/common/proxy_wasm/ngx_proxy_wasm_host.c

```diff
@@ -1807,7 +1834,202 @@ ngx_proxy_wasm_hfuncs_increment_metric(ngx_wavm_instance_t *instance,
 
 
 /* custom extension points */
-/* NYI */
+
+
+#if (NGX_WASM_LUA)
+static void
+ngx_proxy_wasm_hfuncs_resolve_lua_handler(ngx_resolver_ctx_t *rslv_ctx)
+{
+#if (NGX_HAVE_INET6)
+    struct sockaddr_in6          *sin6;
+#endif
+    struct sockaddr_in           *sin;
+    u_char                       *p;
+    ngx_str_t                     args;
+    ngx_wasm_lua_ctx_t           *lctx;
+    ngx_wasm_socket_tcp_t        *sock;
+    ngx_proxy_wasm_foreign_cb_t  *cb;
+    u_short                       sa_family = AF_INET;
+    static size_t                 ipv6_len = 16;
+    u_char                        buf[rslv_ctx->name.len + ipv6_len + 1];
+
+    sock = (ngx_wasm_socket_tcp_t *) rslv_ctx->data;
+    lctx = sock->lctx;
+    cb = (ngx_proxy_wasm_foreign_cb_t *) sock->data;
+    p = buf;
+
+    ngx_memzero(buf, sizeof(buf));
+
+    if (rslv_ctx->state || !rslv_ctx->naddrs) {
+        p++;
+        goto not_found;
+    }
+
+    switch (sa_family) {
+#if (NGX_HAVE_INET6)
+    case AF_INET6:
+          sin6 = (struct sockaddr_in6 *) rslv_ctx->addr.sockaddr;
+
+          *(p++) = sizeof(struct in6_addr);
+          p = ngx_cpymem(p, &sin6->sin6_addr, sizeof(struct in6_addr));
+
+          break;
+#endif
+
+    default: /* AF_INET */
+          sin = (struct sockaddr_in *) rslv_ctx->addr.sockaddr;
+
+          *(p++) = sizeof(struct in_addr);
+          p = ngx_cpymem(p, (u_char *) &sin->sin_addr, sizeof(struct in_addr));
+    }
+
+not_found:
+
+    p = ngx_cpymem(p, rslv_ctx->name.data, rslv_ctx->name.len);
+    args.data = buf;
+    args.len = p - buf;
+
+    if (ngx_proxy_wasm_foreign_callback_buffer_create(cb, args.len) != NGX_OK) {
+        goto error;
+    }
+
+    if (ngx_proxy_wasm_foreign_callback_buffer_write(cb, &args) != NGX_OK) {
+        goto error;
+    }
+
+    if (lctx->yielded) {
+        ngx_proxy_wasm_foreign_callback(cb);
+
+        if (rslv_ctx->state == NGX_WASM_LUA_RESOLVE_ERR) {
+            ngx_wasm_resume(&cb->rctx->env);
+        }
+    }
+
+error:
+
+    ngx_free(rslv_ctx);
+}
+
+
+static ngx_int_t
+ngx_proxy_wasm_hfuncs_resolve_lua(ngx_wavm_instance_t *instance,
+    ngx_http_wasm_req_ctx_t *rctx, ngx_str_t *fargs, ngx_wavm_ptr_t *ret_data,
+    int32_t *ret_size, wasm_val_t rets[])
+{
+    size_t                        s;
+    ngx_int_t                     rc;
+    ngx_buf_t                    *b;
+    ngx_http_request_t           *r;
+    ngx_resolver_ctx_t           *rslv_ctx;
+    ngx_wasm_socket_tcp_t        *sock;
+    ngx_proxy_wasm_exec_t        *pwexec;
+    ngx_proxy_wasm_foreign_cb_t  *cb;
+    ngx_wavm_ptr_t                p = 0;
+
+    pwexec = ngx_proxy_wasm_instance2pwexec(instance);
+
+    cb = ngx_proxy_wasm_foreign_callback_alloc(pwexec);
+    if (cb == NULL) {
+        goto error;
+    }
+
+    cb->fcode = NGX_PROXY_WASM_FOREIGN_RESOLVE;
```

Comment:
Let's try with a function pointer when all those routiens move to "dispatch_call.c"

Reply:
The problem with using a function pointer here is that later we have to provide a function id to the Proxy-Wasm filter when we invoke `proxy_on_foreign_function` indicating the function originally executed by `call_foreign_function`.

As it is, `ngx_proxy_wasm_foreign_call_t` is simply a structure carrying the return value of a function invoked with `call_foreign_function` which yielded before returning.

Reply:
Hmmm ok then, sounds good (I now recall realizing that before our call the other day). Let's keep foreign call ops as an enum then.